Rule SecureAllEndpointsWithScopesRuleTest throws NPE when providing OpenAPI with no scopes defined

I stumbled across an issue when testing an OpenAPI specification without having OAuth2 scopes defined properly. I got a 500 and a NullPointerException within the server application.

```
Cannot submit file for linting. HTTP Status: 500, Response: {"title":"Internal Server Error","status":500,"detail":"check invocation failed: id=105 title=Secure All Endpoints With Scopes checkName=checkOperationsAreScoped reason=java.lang.NullPointerException"}
```

```
java.lang.RuntimeException: check invocation failed: id=105 title=Secure All Endpoints With Scopes checkName=checkOperationsAreScoped reason=java.lang.NullPointerException
        at de.zalando.zally.rule.RulesValidator.invoke(RulesValidator.kt:63) ~[main/:na]
        at de.zalando.zally.rule.RulesValidator.validate(RulesValidator.kt:40) ~[main/:na]
        at de.zalando.zally.rule.CompositeRulesValidator.validate(CompositeRulesValidator.kt:13) ~[main/:na]
        at de.zalando.zally.apireview.ApiViolationsController.validate(ApiViolationsController.kt:47) ~[main/:na]
        ... 127 common frames omitted
Caused by: java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_201]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_201]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_201]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_201]
        at de.zalando.zally.rule.RulesValidator.invoke(RulesValidator.kt:61) ~[main/:na]
        ... 97 common frames omitted
Caused by: java.lang.NullPointerException: null
        at de.zalando.zally.rule.zalando.SecureAllEndpointsWithScopesRule.defined(SecureAllEndpointsWithScopesRule.kt:76) ~[main/:na]
        at de.zalando.zally.rule.zalando.SecureAllEndpointsWithScopesRule.checkOperationsAreScoped(SecureAllEndpointsWithScopesRule.kt:49) ~[main/:na]
        ... 102 common frames omitted
```

I will provide a fix and test for this issue.

--- a/server/src/main/java/de/zalando/zally/rule/zalando/SecureAllEndpointsWithScopesRule.kt
+++ b/server/src/main/java/de/zalando/zally/rule/zalando/SecureAllEndpointsWithScopesRule.kt
@@ -73,7 +73,7 @@ class SecureAllEndpointsWithScopesRule(rulesConfig: Config) {
 
     private fun defined(api: OpenAPI): Map<String, Set<String>> = api.components?.securitySchemes.orEmpty()
         .filterValues { scheme -> scheme.type == SecurityScheme.Type.OAUTH2 }
-        .mapValues { it.value.allFlows().flatMap { it.scopes.keys }.toSet() }
+        .mapValues { it.value.allFlows().flatMap { it.scopes?.keys.orEmpty() }.toSet() }
 
     private fun requested(
         api: OpenAPI,

--- a/server/src/test/java/de/zalando/zally/rule/zalando/SecureAllEndpointsWithScopesRuleTest.kt
+++ b/server/src/test/java/de/zalando/zally/rule/zalando/SecureAllEndpointsWithScopesRuleTest.kt
@@ -1,6 +1,7 @@
 package de.zalando.zally.rule.zalando
 
 import de.zalando.zally.getConfigFromContent
+import de.zalando.zally.getOpenApiContextFromContent
 import de.zalando.zally.getSwaggerContextFromContent
 import de.zalando.zally.rule.ZallyAssertions
 import org.intellij.lang.annotations.Language
@@ -263,5 +264,36 @@ class SecureAllEndpointsWithScopesRuleTest {
         val violations = rule.checkOperationsAreScoped(context)
 
         ZallyAssertions.assertThat(violations).isEmpty()
+    }
+
+    @Test
+    fun `checkOperationsAreScoped with no scopes defined with OpenAPI components`() {
+        @Language("YAML")
+        val yaml = """
+            openapi: 3.0.1
+            
+            paths:
+              '/things':
+                get:
+                  responses:
+                    200:
+                      description: OK
+            
+            components:      
+              securitySchemes:
+                oauth2:
+                  type: oauth2
+                  flows:
+                    clientCredentials:
+                      tokenUrl: 'https://example.com'
+            """.trimIndent()
+
+        val context = getOpenApiContextFromContent(yaml)
+
+        val violations = rule.checkOperationsAreScoped(context)
+
+        ZallyAssertions.assertThat(violations)
+            .descriptionsEqualTo("Endpoint not secured by OAuth2 scope(s)")
+            .pointersEqualTo("/paths/~1things/get")
     }
 }

